exercises/parser_combinators/descr.md

```diff
@@ -0,0 +1,636 @@
+Calling this function, by applying it to the value `()`, amounts to requesting
+the head of the sequence. This head can be either `Nil`, which means that the
+sequence is empty, or `Cons (x, xs)`, which means that the first element of
+the sequence is `x` and the remaining elements form another sequence `xs`. It
+is worth noting that `xs` is itself a function, so the elements of the
+sequence `xs` need not be explicitly computed until `xs` is applied.
+
+Sequences are closely related to *iterators* in object-oriented languages,
+such as C++ and Java. Yet, sequences are much simpler than iterators, for
+two reasons:
+
+* they involve no mutable state;
+
+* they are just as easy to construct and to use as ordinary lists.
+
+The functions `nil`, `cons`, and `singleton` are constructor functions.
+
+The functions `map`, `concat`, `flatten` are analogues for sequences
+of the standard list functions `List.map`, `(@)`, and
+`List.flatten`.
+
+The function `Seq.take` truncates a sequence at a certain length:
+`Seq.take n xs` is a sequence that begins like `xs` but has at most
+`n` elements.
+
+The function `Seq.head` demands the first element of a sequence. If
+the sequence begins with an element `x`, then `Some x` is returned;
+otherwise, `None` is returned. This forces enough computation to
+take place so as to be able to produce the first element of the
+sequence.
+
+The functions `Seq.of_list` and `Seq.to_list` convert between lists
+and sequences, both ways. One must keep in mind that applying
+`Seq.to_list` to a sequence `xs` causes all of its elements to be
+demanded: that is, it forces all of the suspended computations to
+take place. In particular, if `xs` is an infinite sequence, then
+`Seq.to_list xs` does not terminate.
+
+## What's a Parser?
+
+As explained earlier, a parser reads a piece of input text and produces zero,
+one, or more results, where each result represents a possible interpretation
+of this text.
+
+We wish to construct parsers in a compositional way: that is, we wish to
+construct complex parsers by combining simpler parsers. Thus, we must allow
+for the possibility that a parser does not necessarily read *all* of the
+input, from beginning to end, but reads only a *fragment* of the input. In
+other words, if a **cursor** is a pointer into the input text, then a parser
+must receive an initial cursor as an argument. Furthermore, instead of
+returning just a number of *results*, a parser must return a number of *pairs
+of a result and a final cursor*, so that we can tell how far it has
+progressed.
+
+This leads us to the following type definition:
+
+```
+type 'a parser =
+  cursor -> ('a * cursor) NonDet.m
+```
+
+Thus, a parser is a function which, when applied to an initial cursor,
+constructs a nondeterministic computation which, when executed, produces a
+number of pairs of a result and a final cursor.
+
+Now, what is a cursor? It turns out that we need only one operation on
```

Comment:
Don't you also need an operation to get the initial cursor for a given text?

Reply:
Indeed, this is explained later on, where I write: "the input string `input` must be converted to a sequence of characters, therefore, a cursor".